### DISCUSSION TITLE
Discontinuous Galerkin Advection in 3D

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#18210] [Q&A Modules: General] [answered]
Has anybody developed code and examples to use Discontinuous Galerkin in 3D, for instance to model advection?

I was looking at the test case here:
https://github.com/idaholab/moose/blob/next/modules/rdg/test/tests/advection_1d/1d_aefv_square_wave.i

My understanding is that if I need to implement the same model in 3D model,
I will need to:

1) Extend AEFVSlopeLimitingOneD to calculate gradients in 3D
2) Extend AEFVUpwindInternalSideFlux to use a 3D velocity vector

while the AEFVKernel seems to be already suitable for 3D, 
and same for AEFVMaterial.

Am I right?
Could you give me some hints about the implementation?

Thank you very much in advance for your suggestions,
Best Regards,
Nicolò

**Comment by jbadger95** [comment]
Disclaimer, I'm fairly new to MOOSE and am still learning but I think the examples you're looking at are more about using DG to implement a Finite Volume method (more natural FV capabilities are in the works) and less about using DG to solve the problem. In the tests directory you'll find a lot more examples with DG (for example 3D DG examples). Checkout test/tests/dgkernels, I haven't looked at them in depth but just glancing I'm seeing 3d_diffusion, ad_dg_convection, advection_diffusion_mixed_bcs.

**Reply by ngrilli** [reply]
Dear @jbadger95 

Thank you very much for pointing me to those files.
Now I see DGKernels are also implemented in the framework and not only in the rdg module.
I will try them. Thank you again.

Best Regards,
Nicolò

**Comment by ngrilli** [comment]
Dear MOOSE users,

following up my initial question:
now I learned how to use the DG kernel for advection.

Now I would like to do coupled advection, I mean I want to implement the equation:

![immagine](https://user-images.githubusercontent.com/22821173/124002802-17c1cf80-d9ce-11eb-88b6-50b94ed42c6e.png)

where u is my variable, rho is a coupled variable and v the velocity.

The question is: how should I modify the DGConvection kernel?

The main challenge is that I need to get the coupled variable in the neighbouring element,
so, for instance in the residual calculation, I will have a variable _rho_coupled_neighbor 
instead of the variable _u_neighbor

while I think the variable _u[_qp] can be simply substituted by _rho_coupled[_qp] which is obtained by:
_rho_coupled(coupledValue("rho_coupled")), this is the easy part I think. 

I saw in the base class DGKernel that the variable in the neighbour element is obtained by:
_u_neighbor(_is_implicit ? _var.slnNeighbor() : _var.slnOldNeighbor()),

However: _var(*mooseVariable()) gets the variable on which the kernel is applied, so u.

So, the question is: is there a function similar to mooseVariable() which can provide me a coupled variable instead, given its name?

Any other suggestion to reach my goal?

**Reply by YaqiWang** [reply]
`DGCoupledDiffusion.C` under moose/test/src/dgkernels shows us how to couple variables in DG kernels.

**Reply by YaqiWang** [reply]
It is not using more strait-forward functions provided in `NeighborCoupleable`. Maybe MOOSE developers can comment on this? I was expecting to see something like
```
    _v(coupledValue("v")),
    _v_neighbor(coupledNeighborValue("v")),
```
You can try this to see if it works or not.

**Comment by ngrilli** [comment, marked as answer]
Dear MOOSE users,

following up my initial question:
now I have implemented the DG kernel for coupled advection:
https://github.com/ngrilli/c_pfor_am/blob/main/src/dgkernels/DGAdvectionCoupledVConst.C

The equations I am trying to solve are:
![immagine](https://user-images.githubusercontent.com/22821173/124304281-62bd1d80-db5b-11eb-8d0a-38504a0b5f2d.png)

$\rho_e$ and $\rho_t$ are the two variables.
v is a constant velocity (v = 1.0) along positive x axis.
An initial vertical strip contains the non-zero phase field and I want to see the advection towards the right.

I think the implementation is correct because I see the correct solution at up to time t = 1.0 when the variable reaches at the boundary,
see image below:
![immagine](https://user-images.githubusercontent.com/22821173/124305723-5f2a9600-db5d-11eb-8948-33c5770d7fcd.png)

I think the advection s correct, but there are two things:
1) Periodic BC do not work, the field clearly does not transfer from right to left
2) After a certain time, when the field is at the boundary, I see $\rho_e$ and $\rho_t$ get very large value and simulation diverge.

Point 2) I really don't understand because there is nothing in the equations themselves that can give this divergence.

I tried with different shape functions (MONOMIAL, L2_LAGRANGE)
and with/without periodic BC, Dirichlet BC also give the same effect.

Any suggestion is very much appreciated.
The input file is copied in the following, and also available here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/DGKernels/DGtestVConst.i

Best Regards,
Nicolò

```
[Mesh]

  [./gen]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 10
    ny = 10
    nz = 1
    xmax = 2.0
    ymax = 2.0
    zmax = 0.2
    elem_type = HEX8
    displacements = 'disp_x disp_y disp_z'  
  [../] 

[]

[GlobalParams]

[]

[Variables]
 
  [./rho_t]
    order = CONSTANT
    family = MONOMIAL
    [./InitialCondition]
      type = FunctionIC
      function = init_rho_t
    [../]
  [../]
  
  [./rho_gnd_edge]
    order = CONSTANT
    family = MONOMIAL
    [./InitialCondition]
      type = FunctionIC
      function = init_rho_gnd_edge
    [../]
  [../]
  
[]

[AuxVariables]

[]

[Functions]

  [./dts]
    type = PiecewiseConstant
    x = '0.0 10.0'
    y = '0.1 0.1'
  [../]
  
  [./init_rho_gnd_edge]
    type = ParsedFunction
	value = 'if(3.0*(x-1.0),0.0,1.0)'
  [../]
  
  [./init_rho_t]
    type = ParsedFunction
	value = 'if(3.0*(x-1.0),0.0,1.0)'
  [../]
  
[]

[UserObjects]

[]

[Kernels]

  [./drho_gnd_edge_dt]
    type = TimeDerivative
    variable = rho_gnd_edge
  [../]

  [./drho_t_dt]
    type = TimeDerivative
    variable = rho_t
  [../]  
    
[]

[DGKernels]

  [./rho_t_advection_edge]
    type = DGAdvectionCoupledVConst
    variable = rho_gnd_edge
	rho_coupled = rho_t
    velocity = '1 0 0'
  [../] 
  
  [./rho_gnd_edge_advection]
    type = DGAdvectionCoupledVConst
    variable = rho_t
	rho_coupled = rho_gnd_edge
    velocity = '1 0 0'
  [../]

[]

[AuxKernels]

[]

[BCs]
  
  [./Periodic]
  
    [./auto_rho_t_boundary_x]
      variable = rho_t
      primary = 'left'
	  secondary = 'right'
	  translation = '2.0 0.0 0.0'
    [../]
    [./auto_rho_gnd_edge_boundary_x]
      variable = rho_gnd_edge
      primary = 'left'
	  secondary = 'right'
      translation = '2.0 0.0 0.0'  
    [../]
	
    [./auto_rho_t_boundary_y]
      variable = rho_t
      primary = 'bottom'
	  secondary = 'top'
	  translation = '0.0 2.0 0.0'
    [../]
    [./auto_rho_gnd_edge_boundary_y]
      variable = rho_gnd_edge
      primary = 'bottom'
	  secondary = 'top'
      translation = '0.0 2.0 0.0'	  
    [../]
	
    [./auto_rho_t_boundary_z]
      variable = rho_t
      primary = 'back'
	  secondary = 'front'
	  translation = '0.0 0.0 0.2'
    [../]
    [./auto_rho_gnd_edge_boundary_z]
      variable = rho_gnd_edge
      primary = 'back'
	  secondary = 'front'
	  translation = '0.0 0.0 0.2'
    [../]

  [../]  
  
[]
 
[Postprocessors]

[]

[Materials]

[]

[Preconditioning]
  
  active = 'smp'
  
  [./smp]
    type = SMP
    full = true
  [../]
  
[]

[Executioner]

  type = Transient
  solve_type = 'PJFNK'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu superlu_dist'
  
  line_search = 'none'
  automatic_scaling = true
  
  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  
  [./TimeStepper]
    type = FunctionDTGrowth
    function = dts
	cutback_factor_at_failure = 0.1
	growth_factor = 1.2
  [../]

  start_time = 0.0
  end_time = 10.0
  
  dtmin = 1.0e-10
  timestep_tolerance = 1.0e-10
[]

[Outputs]

  csv = false

  [./out]
    type = Exodus
    interval = 1
  [../]
  
[]
```

**Reply by ngrilli** [reply]
@YaqiWang maybe you have a clue about this?
Thank you
Nicolò

**Reply by GiudGiud** [reply]
does it work with reflective?
There's definitely some gaps in the implementation of periodic BCs
@gridley has looked at that recently, maybe he can help here

**Reply by ngrilli** [reply]
Dear @GiudGiud 

Apologies but I am not sure what you mean by reflective BC.
Could you point me to an example in moose?

I only tried removing all BC, so it should be zero flux and I get the same solution as with periodic.
I also tried dirichlet and penalty dirichlet but I get the same problem of the high value of the variable and oscillations appearing
at the boundary after some time.

I don't specifically need periodic BC, but I was just trying to setup a simulation
that can work even if there are accumulation of the variable at the boundaries.

Otherwise, is there any other way to make conservative advection with a coupled variable in moose?
I made this code some time ago:
https://github.com/ngrilli/c_pfor_am/blob/main/src/kernels/ConservativeAdvectionCoupled.C

but it was not working well, I am not enough familiar with the framework to modify properly
the function fullUpwind(JacRes res_or_jac)
in such a way that it can work with a coupled variable.

This code was taken from:
https://github.com/idaholab/moose/blob/next/framework/src/kernels/ConservativeAdvection.C

but my modifications of fullUpwind(JacRes res_or_jac) were probably not correct.

If someone can help with this, it would be great.

Nicolò

**Reply by GiudGiud** [reply]
Reflective should be the zero flux here. 

The other way, and probably the best way, to do conservative advection is to use finite volume. You can have a look at the advection kernels for incompressible navier stokes. For advection with a constant velocity there's a simple one in the framework, FVAdvection. 

For advection with a coupled variable velocity, INSFVMomentum/MassAdvection are more useful. 

Note that DG should work too, Im just less familiar with it so cant really help

Guillaume

**Reply by YaqiWang** [reply]
Sorry, I did not notice this. You seem missing a kernel for the convection term. Check this `moose/test/tests//dgkernels/1d_advection_dg/1d_advection_dg.i` out. We apply integration by parts for the convection term multiplied with a test function resulting into a normal kernel and a dg kernel.

**Reply by ngrilli** [reply]
Dear @YaqiWang 

Thank you, really, your comment was illuminating for me.
So let me see if I understand well:
1) The DG kernels are taking care of the discontinuity of the shape functions at the elements boundaries.
2) The normal kernels are taking care of the bulk term, where shape functions can have a gradient, which must be considered for advection.
Therefore, I need to use both at the same time.
Does this make sense?

I have implemented both normal kernel for coupled advection and DG kernel for coupled advection and I am using them together.
The result is correct now, see the following picture compared to the one I posted above:
![immagine](https://user-images.githubusercontent.com/22821173/124673463-c6737d80-deb0-11eb-9ebf-a0d938419e58.png)

I can see that the advection is definitely conservative and there is no more solution divergence when the
field remains blocked at the boundary like before.

Main features:
1) I am able to run these simulations with both first order monomial and constant monomial variables
2) periodic BC do not do anything in both cases, but that was the same as I found earlier, not very important for my project anyway
3) the key to avoid the divergence at the end is actually to add the line "implicit = false" in all the advection kernels,
as it is done in the example in moose:
https://github.com/idaholab/moose/tree/next/test/tests/dgkernels/1d_advection_dg
If I don't use "implicit = false", I have the divergence problem discussed above even with the combined DG + non-DG kernels.

Anyway the solution looks much more correct now and I will pursue further this method for my problem.
If anyone is interested, my kernels for coupled advection and test cases are here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/DGKernels/DGtestConstMonomial.i
https://github.com/ngrilli/c_pfor_am/blob/main/src/dgkernels/DGAdvectionCoupledVConst.C
https://github.com/ngrilli/c_pfor_am/blob/main/src/kernels/AdvectionCoupledVConst.C

Thank you very much again,
Nicolò

**Reply by YaqiWang** [reply]
Glad it is working. I am not so sure about this `implicit = false`. This is making the kernel coupling the variable of the previous time step. I would think it should work without this `implicit`. You may want to double check.
